packages/teraslice/lib/cluster/services/cluster/backends/kubernetes/k8s.js

```
@@ -261,18 +299,48 @@ class K8s {
 
     /**
      * Delete all of the deployments and services related to the specified exId
+     *
+     * The process here waits for the worker pods to completely exit before
+     * terminating the execution controller pod.  The intent is to avoid having
+     * a worker timeout when it tries to tell the execution controller it is
+     * exiting.
+     *
      * @param  {String}  exId ID of the execution
      * @return {Promise}
      */
     async deleteExecution(exId) {
+        const r = [];
         if (!exId) {
             throw new Error('deleteExecution requires an executionId');
         }
 
-        return Promise.all([
-            this._deleteObjByExId(exId, 'worker', 'deployments'),
-            this._deleteObjByExId(exId, 'execution_controller', 'jobs'),
-        ]);
+        try {
+            this.logger.info(`Deleting worker deployment for ex_id: ${exId}`);
+            r.push(await this._deleteObjByExId(exId, 'worker', 'deployments'));
+        } catch (e) {
+            const err = new Error(`Error deleting workers: ${e}`);
+            this.logger.error(err);
+            return Promise.reject(err);
+        }
+
+        await this.waitForNumPods(
+            0,
+            `app.kubernetes.io/component=worker,teraslice.terascope.io/exId=${exId}`,
+            null,
+            600000 // 10 minutes, I wanted this to be longer than the pod shutdown timeout
```

The value of `600000` is arbitrary but designed to be longer than the pod shutdown timeout.  Maybe it should use the pod shutdown timeout and add like 30s to it.  Also, I guess this should probably be in a try/catch ... with the catch doing ... `Promise.reject`?

---

Using a try catch around `waitForNumPods` will ignore all errors in that function, it would be safer to add a try catch around just the [k8s request](https://github.com/terascope/teraslice/blob/43a7b182827c7038dd3a5168641ff5750f223702/packages/teraslice/lib/cluster/services/cluster/backends/kubernetes/k8s.js#L118-L119).

---

I've rethought this behavior ... the critical thing that I want to happen here is that in the event of an error, I want the execution controller to be deleted.  Or at the very least, I want the code to try and do it.  Even if the worker delete fails.

The k8s worker deployment now has an `ownerReference` to the execution controller job such that if the job is deleted, so is the deployment (all the pods already link to the job/deployment).  So I rely on k8s garbage collection when the code encounters an unretryable error.